tools/Dashboard/DevHome.Dashboard/ComSafeWidgetObjects/ComSafeWidgetDefinition.cs

```diff
@@ -0,0 +1,189 @@
+            }
+            catch (COMException ex) when (ex.HResult == RpcServerUnavailable || ex.HResult == RpcCallFailed)
+            {
+                _log.Warning(ex, $"Failed to operate on out-of-proc object with error code: 0x{ex.HResult:x}");
+                await GetNewOopWidgetDefinitionAsync();
+            }
+            catch (Exception ex)
+            {
+                _log.Error(ex, "Exception getting widget capabilities from widget:");
+                return null;
+            }
+        }
+
+        return null;
+    }
+
+    private async Task GetNewOopWidgetDefinitionAsync()
+    {
+        _oopWidgetDefinition = null;
+        await LazilyLoadOopWidgetDefinition();
+    }
+
+    private async Task LazilyLoadOopWidgetDefinition()
+    {
+        var attempt = 0;
+        while (attempt++ < 3 && (_oopWidgetDefinition == null || _hasValidProperties == false))
+        {
+            try
+            {
+                _oopWidgetDefinition ??= await Application.Current.GetService<IWidgetHostingService>().GetWidgetDefinitionAsync(Id);
+
+                if (!_hasValidProperties)
+                {
+                    await Task.Run(() =>
+                    {
+                        AllowMultiple = _oopWidgetDefinition.AllowMultiple;
+                        Description = _oopWidgetDefinition.Description;
+                        DisplayTitle = _oopWidgetDefinition.DisplayTitle;
+                        Id = _oopWidgetDefinition.Id;
+                        ProviderDefinition = _oopWidgetDefinition.ProviderDefinition;
+                        ProviderDefinitionDisplayName = _oopWidgetDefinition.ProviderDefinition.DisplayName;
+                        ProviderDefinitionId = _oopWidgetDefinition.ProviderDefinition.Id;
+                        AdditionalInfoUri = _oopWidgetDefinition.AdditionalInfoUri;
+                        IsCustomizable = _oopWidgetDefinition.IsCustomizable;
+                        Type = _oopWidgetDefinition.Type;
+                        _hasValidProperties = true;
+                    });
+                }
+            }
+            catch (Exception ex)
+            {
+                _log.Warning(ex, "Failed to get properties of out-of-proc object");
+            }
+        }
+    }
+
+    /// <summary>
+    /// Get a WidgetDefinition's ID from a WidgetDefinition object. Tries multiple times in case of COM exceptions.
+    /// </summary>
+    /// <param name="widgetDefinition">WidgetDefinition</param>
+    /// <returns>The Widget's Id, or in the case of failure string.Empty</returns>
+    public static async Task<string> GetIdFromUnsafeWidgetDefinitionAsync(WidgetDefinition widgetDefinition)
+    {
+        var retries = 5;
+
+        return await Task.Run(() =>
+        {
+            while (retries-- > 0)
+            {
+                try
+                {
+                    return widgetDefinition.Id;

```

Comment:
We probably don't need to retry here because if we get an RPC error the first time then we will continue getting an error until a new instance of `widgetDefinition` is created.